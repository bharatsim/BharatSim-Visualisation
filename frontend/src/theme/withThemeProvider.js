import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
