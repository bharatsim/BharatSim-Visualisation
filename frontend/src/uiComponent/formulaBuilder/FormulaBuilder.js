import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { Box, Button, FormHelperText, Typography } from '@material-ui/core';
import { highlight, languages } from 'prismjs';
import Editor from 'react-simple-code-editor';
import 'prismjs/components/prism-javascript';
import { withStyles } from '@material-ui/core/styles';
import clsx from 'clsx';
import { parse } from 'mathjs';

import { useStyles } from './formulaBuilderStyles';
import List from '../List';
import './prismTheme.css';

const OperatorButton = withStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: theme.spacing(8),
    width: theme.spacing(8),
    boxShadow: 'unset',
    backgroundColor: theme.colors.grayScale[100],
    padding: 0,
    minWidth: 'unset',
    ...theme.typography.subtitle1,
  },
}))(Button);

const EDITOR_TEXTAREA = '.npm__react-simple-code-editor__textarea';

const typeOfCode = {
  OPERATOR: 'operator',
  FIELDS: 'fields',
};

function getCursorPosition(curCursorPos, stringToAdd) {
  return curCursorPos + stringToAdd.length;
}

function insertAtIndex(str, index, stringToAdd) {
  return str.substring(0, index) + stringToAdd + str.substring(index, str.length);
}

function setCursor(textArea, position) {
  setTimeout(() => {
    textArea.setSelectionRange(position, position);
  }, 1);
}

function parseError(error) {
  const errorType = error.name;
  let errorCharPosition = error.message.search(/\(char [0-9]*\)/g);
  errorCharPosition = errorCharPosition < 0 ? undefined : errorCharPosition;
  return `${errorType}: ${error.message.slice(0, errorCharPosition)}`;
}

function getFormattedCode(code, type) {
  return {
    [typeOfCode.FIELDS]: ` "${code}" `,
    [typeOfCode.OPERATOR]: ` ${code} `,
  }[type];
}

function FormulaBuilder({ fields, operators, buttonLabel, onClick, defaultExpression }) {
  const classes = useStyles();
  const [expression, setExpression] = useState(defaultExpression);
  const [error, setError] = useState('');

  const ref = useRef();
  useEffect(() => {
    ref.current = document.querySelector(EDITOR_TEXTAREA);
  }, []);

  function handleCalculateFormula() {
    try {
      if (!expression.trim()) {
        throw new Error('Expression can not be empty');
      }
      const parsedExpression = parse(expression);
      onClick(parsedExpression, expression);
    } catch (e) {
      setError(parseError(e));
    }
  }

  function handleChange(code) {
    if (error) setError('');
    setExpression(code);
  }

  function handleClick(code, type) {
    ref.current.focus();
    if (error) setError('');
    const cursorPos = ref.current.selectionStart;
    const codeToInsert = getFormattedCode(code, type);
    setExpression(insertAtIndex(expression, cursorPos, codeToInsert));
    const newCursorPosition = getCursorPosition(cursorPos, codeToInsert);
    setCursor(ref.current, newCursorPosition);
  }

  function digitAndOperatorOnly(evt) {
    const charCode = evt.which;
    const isNumber = charCode >= 48 && charCode <= 57;
    const isOperator = operators.includes(evt.key);
    const isSpace = charCode === 32;
    if (isOperator || isNumber || isSpace) {
      return;
    }
    evt.preventDefault();
  }

  return (
    <Box>
      <Box className={classes.mainContainer}>
        <Box className={clsx(classes.container, classes.containerEditor)}>
          <Typography variant="subtitle1">Formula</Typography>
          <Editor
            placeholder="Type your formulaâ€¦"
            value={expression}
            onValueChange={handleChange}
            onKeyPress={digitAndOperatorOnly}
            highlight={(code) => highlight(code, languages.js)}
            padding={10}
            className={classes.editor}
            data-testid="textarea"
          />
          <Box className={classes.errorContainer}>
            {!!error && (
              <FormHelperText error classes={{ root: classes.helperText }}>
                {error}
              </FormHelperText>
            )}
          </Box>
        </Box>
        <Box className={clsx(classes.container, classes.containerFunctions)}>
          <Typography variant="subtitle1">Fields</Typography>
          <List listItem={fields} onClick={(field) => handleClick(field, typeOfCode.FIELDS)} />
        </Box>
      </Box>
      <Box className={classes.actionContainer}>
        <Box className={classes.operatorContainer}>
          {operators.map((operator) => (
            <OperatorButton
              onClick={() => handleClick(operator, typeOfCode.OPERATOR)}
              key={operator}
            >
              {operator}
            </OperatorButton>
          ))}
        </Box>
        <Button size="small" variant="contained" color="primary" onClick={handleCalculateFormula}>
          {buttonLabel}
        </Button>
      </Box>
    </Box>
  );
}

FormulaBuilder.propTypes = {
  fields: PropTypes.arrayOf(PropTypes.string).isRequired,
  operators: PropTypes.arrayOf(PropTypes.string).isRequired,
  buttonLabel: PropTypes.string,
  defaultExpression: PropTypes.string,
  onClick: PropTypes.func.isRequired,
};

FormulaBuilder.defaultProps = {
  buttonLabel: 'Calculate',
  defaultExpression: '',
};

export default FormulaBuilder;
