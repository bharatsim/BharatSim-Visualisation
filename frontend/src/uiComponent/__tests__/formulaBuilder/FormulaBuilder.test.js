import React from 'react';
import { render } from '@testing-library/react';
import { fireEvent } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import FormulaBuilder from '../../formulaBuilder/FormulaBuilder';
import withThemeProvider from '../../../theme/withThemeProvider';

const operator = ['+', '-', '*'];
const fields = ['field1', 'field2'];

const FormulaBuilderWithProviders = withThemeProvider(FormulaBuilder);

describe('<FormulaBuilder />', () => {
  it('should match snapshot of component', () => {
    const { container } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={jest.fn()}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onClick callback with entered formula', () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const field1Button = getByText('field1');
    const plusButton = getByText('+');
    const calculateButton = getByText('calculate');

    fireEvent.click(field1Button);
    fireEvent.click(plusButton);
    fireEvent.click(field1Button);
    fireEvent.click(calculateButton);

    expect(onClick).toHaveBeenCalledWith(' "field1"  +  "field1" ');
  });

  it('should through error if formula is wrong', () => {
    const onClick = jest.fn();
    const { getByText, queryByText } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const field1Button = getByText('field1');
    const calculateButton = getByText('calculate');
    const plusButton = getByText('+');

    fireEvent.click(field1Button);
    fireEvent.click(field1Button);
    fireEvent.click(plusButton);
    fireEvent.click(calculateButton);

    expect(queryByText('SyntaxError: Unexpected operator "')).not.toBeNull();
  });

  it('should through error if formula is empty', () => {
    const onClick = jest.fn();
    const { getByText, queryByText } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const calculateButton = getByText('calculate');
    fireEvent.click(calculateButton);

    expect(queryByText('Error: Expression can not be empty')).not.toBeNull();
  });

  it('should not allow other digit, opertor and space to type', async () => {
    const onClick = jest.fn();
    render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const textArea = document.querySelector('.npm__react-simple-code-editor__textarea');

    userEvent.type(textArea, 'ab=}|@');

    expect(textArea).toHaveValue('');
  });

  it('should allow digit, operator and space to type', async () => {
    const onClick = jest.fn();
    render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const textArea = document.querySelector('.npm__react-simple-code-editor__textarea');

    await userEvent.type(textArea, '12 + 13');

    expect(textArea).toHaveValue('12 + 13');
  });

  it('should remove error on change of expression', () => {
    const onClick = jest.fn();
    const { getByText, queryByText } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const calculateButton = getByText('calculate');
    fireEvent.click(calculateButton);
    const textArea = document.querySelector('.npm__react-simple-code-editor__textarea');

    expect(queryByText('Error: Expression can not be empty')).not.toBeNull();
    userEvent.type(textArea, ' 12 + 13');
    expect(queryByText('Error: Expression can not be empty')).toBeNull();
  });

  it('should remove error on click of any field', () => {
    const onClick = jest.fn();
    const { getByText, queryByText } = render(
      <FormulaBuilderWithProviders
        operators={operator}
        onClick={onClick}
        fields={fields}
        buttonLabel="calculate"
      />,
    );

    const calculateButton = getByText('calculate');
    fireEvent.click(calculateButton);

    expect(queryByText('Error: Expression can not be empty')).not.toBeNull();
    const field1Button = getByText('field1');
    fireEvent.click(field1Button);
    expect(queryByText('Error: Expression can not be empty')).toBeNull();
  });
});
