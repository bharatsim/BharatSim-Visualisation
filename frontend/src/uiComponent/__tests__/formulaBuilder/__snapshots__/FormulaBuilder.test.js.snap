// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormulaBuilder /> should match snapshot of component 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-11"
  >
    <div
      class="MuiBox-root MuiBox-root-12 makeStyles-mainContainer-1"
    >
      <div
        class="MuiBox-root MuiBox-root-13 makeStyles-container-2 makeStyles-containerEditor-3"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle1"
        >
          Formula
        </h6>
        <div
          class="makeStyles-editor-4"
          data-testid="textarea"
          style="position: relative; text-align: left; box-sizing: border-box; padding: 0px; overflow: hidden;"
        >
          <textarea
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="npm__react-simple-code-editor__textarea"
            data-gramm="false"
            placeholder="Type your formulaâ€¦"
            spellcheck="false"
            style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; resize: none; overflow: hidden; padding: 10px 10px 10px 10px;"
          />
          <pre
            aria-hidden="true"
            style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: relative; pointer-events: none; padding: 10px 10px 10px 10px;"
          >
            <br />
          </pre>
          <style
            type="text/css"
          >
            
/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * Hack to apply on some CSS on IE10 and IE11
 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  /**
    * IE doesn't support '-webkit-text-fill-color'
    * So we use 'color: transparent' to make the text transparent on IE
    * Unlike other browsers, it doesn't affect caret color in IE
    */
  .npm__react-simple-code-editor__textarea {
    color: transparent !important;
  }

  .npm__react-simple-code-editor__textarea::selection {
    background-color: #accef7 !important;
    color: transparent !important;
  }
}

          </style>
        </div>
        <div
          class="MuiBox-root MuiBox-root-14 makeStyles-errorContainer-8"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-15 makeStyles-container-2 makeStyles-containerFunctions-6"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle1"
        >
          Fields
        </h6>
        <div
          class="MuiBox-root MuiBox-root-19 makeStyles-mainContainer-16"
        >
          <ul
            class="MuiList-root makeStyles-root-18 MuiList-padding"
          >
            <li
              class="MuiListItem-root makeStyles-root-17 MuiListItem-gutters"
            >
              field1
            </li>
            <li
              class="MuiListItem-root makeStyles-root-17 MuiListItem-gutters"
            >
              field2
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-20 makeStyles-actionContainer-10"
    >
      <div
        class="MuiBox-root MuiBox-root-21 makeStyles-operatorContainer-5"
      >
        <button
          class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-22 MuiButton-text"
          tabindex="0"
          title="Addition"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            +
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-22 MuiButton-text"
          tabindex="0"
          title="Subtraction"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            -
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-22 MuiButton-text"
          tabindex="0"
          title="Multiplication"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            *
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          calculate
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
